keps/sig-apps/2185-random-pod-select-on-replicaset-downscale/README.md

```diff
@@ -0,0 +1,437 @@
+  - [Scalability](#scalability)
+  - [Troubleshooting](#troubleshooting)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Infrastructure Needed (Optional)](#infrastructure-needed-optional)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [x] (R) KEP approvers have approved the KEP status as `implementable`
+- [x] (R) Design details are appropriately documented
+- [x] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [x] (R) Graduation criteria is in place
+- [x] (R) Production readiness review completed
+- [x] Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The current downscaling algorithm for ReplicaSets prefers to delete Pods that
+have been running for the least amount of time. This heuristic attempts to
+minimize disruption under the premise that newer Pods are likely to be serving a
+lesser amount of clients. However, the heuristic can be detrimental to high
+availability requirements: when a ReplicaSet lands in an imbalanced state across
+failure domains, the heuristic tends to preserve the imbalance after repeated
+up and down scales.
+
+We propose a randomized approach to the downscale Pod victim selection algorithm
+of the ReplicaSet controller to mitigate ReplicaSet imbalance across failure
+domains.
+
+## Motivation
+
+There are scenarios where a ReplicaSet can reach an imbalanced state across
+failure domains. See [user stories](#user-stories) to see one such scenario
+and how a randomized approach solves the issue.
+
+### Goals
+
+- A randomized algorithm for Pod selection when downscaling ReplicaSets.
+- Softly honor the heuristic that prefers to downscale newer Pods first.
+- Validate that the approach is able to get a ReplicaSet out of an imbalanced
+  state.
+
+### Non-Goals
+
+- Provide guarantees of preserving balance during scale down. This introduces
+  a violation of separation of concerns between the ReplicaSet controller and
+  kube-scheduler.
+- Preserve the existing behavior that always downscales the newer Pods first.
+  This order was never guaranteed in the API or user documentation.
+
+## Proposal
+
+
+### User Stories
+
+#### Story 1
```

Comment:
focusing on the upgrade story is more convincing in my opinion

Reply:
True, this story could probably just be shortened to the last 2 steps (5,6): where a new domain is added and upscaled to 3N, containing all the youngest pods. Then downscaled to 2N and all the pods from the new domain are removed. Is that what you mean?

Reply:
Updated to just focus on an upgrade, please check my wording :)

Reply:
Not sure if that's what Abdullah meant, but thanks for the simplification.

For completion, I would say:

A deployment could become imbalanced after:

- An entire failure domain fails and becomes unavailable
- A failure domain gets teared down for maintenance or upgrade
- A new failure domain is added to a cluster

The imbalance cycle goes like follows:

Reply:
This sounds reasonable, but the upgrade case I had in mind is not necessarily about adding a new failure domain, but about a rolling upgrade of nodes: consider a region with three zones, and the nodes get upgraded one zone at a time. As aldo mentioned, you can say that all those cases can lead to imbalance.